### PR TITLE
Added unit tests to cover passing nullable types to System.Diagnostics.Tracing

### PR DESCRIPTION
**Pull request by jbhensley**
This goes with [coreclr PR 20350](https://github.com/dotnet/coreclr/pull/20350). Telemetry data from the App Insights team showed errors being generated from nullable type usage in conjunction with System.Diagnostics.Tracing. An issue was filed as corefx #32606. The coreclr PR fixes the problem and these tests confirm that. Note that these new tests will fail until the coreclr PR is merged and synced. 

Of note: there are two tests already failing:

- `BasicEventSourceTests.TestsUserErrors.Test_BadEventSource_MismatchedIds_WithEtwListener`
- `BasicEventSourceTests.TestsUserErrors.Test_BadEventSource_MismatchedIds`

Those two tests have an open issue (#29754) and should be unrelated.

### CODE REVIEW

File: src/System.Diagnostics.Tracing/tests/BasicEventSourceTest/TestsWrite.cs [hunk 1]
```diff
@@ -513,5 +573,16 @@ private void GenerateArrayTest<T>(ref List<SubTest> tests, EventSource logger, T
             }
             return ret;
         }
+
+        [Fact]
+        public void Test_Write_Nullable_Does_Not_Throw()
```
[the remark this answers is not in the document]
Comment (jbhensley) [reply src/System.Diagnostics.Tracing/tests/BasicEventSourceTest/TestsWrite.cs]: It doesn't. I tossed it in at the beginning while tracking down the problem and didn't take it out. Will remove.

**Comment by jbhensley** [comment]
For the listener, adding a method to `EventSourceTest` that takes and writes `int?` results in "Unsupported type Nullable in event source" coming from `ManifestBuilder` [at this location](https://github.com/dotnet/coreclr/blob/09cc49e8cac72915b72240c766e25ada171e9fe7/src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventSource_CoreCLR.cs#L188). For what it's worth, the `TypeCode` of a nullable is `Object`.

Ultimately, the head of the call stack leads back to `EventListener.EnableEvents`, but this is really `EventSource.CreateManifestAndDescriptors` doing the work of building the manifest.

For reference, I was adding this to `EventSourceTest`:

```C#
[Event(53, Keywords = Keywords.HasNullableArgs, Level = EventLevel.Informational)]
public void EventNullableInt1(int? arg1) { WriteEvent(53, arg1); }
```

and this to `TestsWriteEventToListener.Test_WriteEvent_ArgsBasicTypes`:

```C#
log.EventNullableInt1(10);
Assert.Equal(53, LoudListener.t_lastEvent.EventId);
Assert.Equal(1, LoudListener.t_lastEvent.Payload.Count);
Assert.Equal(10, (int?)LoudListener.t_lastEvent.Payload[0]);
```
as a start, with the intention of adding null values and a mixture of null/non-null. 

All of this code is new to me, so it's taking a bit to get orientated. I'm guessing we need to fix manifest generation to support nullables. I'm just not sure what type string to return because I'm not clear on how the manifest is used. Suggestions?

**Comment by noahfalk** [comment]
Aha, excellent, you've likely discovered another lurking bug because our testing for Nullable was insufficient. Since it will take some additional time and probably a new CoreCLR PR to investigate and fix this second issue I'd suggest we start getting some of this checked in. To do this you could disable or comment out the test edits you made for the EventListener portion and add a comment saying the issue is being tracked by #32796 which I just created. That should get everything passing.

> All of this code is new to me, so it's taking a bit to get orientated. I'm guessing we need to fix manifest generation to support nullables. I'm just not sure what type string to return because I'm not clear on how the manifest is used. Suggestions?

Not to worry, I've only very recently been asked to take ownership of this code at Microsoft so I am in the same boat as you. We can learn together : ) I'm going to sync up to your set of changes so that I can reproduce the same issue and help debug.

**Comment by jbhensley** [comment]
Hmmm. Got a `java.nio.channels.ClosedChannelException` on OSX x64 Debug Build already. Is something up with the CI?

**Comment by noahfalk** [comment]
Sadly my understanding is that this is a non-deterministic bug in Jenkins. When I hit it I ask dotnet-bot to re-run the leg and it (usually) doesn't repro again. You can do that by writing in a comment:

@dotnet-bot test OSX x64 Debug Build

We won't be able to check in these tests until the CoreCLR change has been inserted, so I am expecting the CI to fail right now.

**Comment by jbhensley** [comment]
Thanks. Saw the same error on the CoreCLR PR as well and wasn't sure if that was something that normally happens. Sounds like the CI is a bit buggy.

Looks like the fails coming back now are the ones we except.

**Comment by noahfalk** [comment]
> I'm just not sure what type string to return because I'm not clear on how the manifest is used. Suggestions?

@jbhensley - I put some debugging notes in that issue that hopefully points in the right direction. My understanding of the code is still spotty, but it is improving.

**Comment by jbhensley** [comment]
I'm struggling with some weirdness where tests are failing remote, but not repro'ing locally. Example:

```C#
logger.Write("Int?-12", new { nInteger = nullableInt });
...
Assert.Equal("Int?-12", evt.EventName);
```

passed locally. CI showed this:

error

Like the dash/minus got changed to underscore. I changed the code to this and now it passes:

```C#
logger.Write("Int12", new { nInteger = nullableInt });
...
Assert.Equal("Int12", evt.EventName);
```

Now CI tossing here with `System.InvalidCastException : Unable to cast object of type 'StructValue' to type 'System.Nullable`1[System.Int32]'`:

```C#
var eventNullableInt = (int?)evt.PayloadValue(0, "nInteger");
Assert.Equal(eventNullableInt, nullableInt);
```

This passes locally. I know my local tests are firing, because adding `Assert.Equal(0,1)` right after these lines cause the test to fail. I'm kind of flying blind with different results local vs remote. Fetching latest master and starting a brand new branch to ensure I'm in parity made no difference. Ever seen anything like this?

**Comment by jbhensley** [comment]
Ok, I'm baffled. Current version dumps the type name of `eventNullableInt` to console.

```C#
var eventNullableInt = evt.PayloadValue(0, "nInteger");
Console.WriteLine($"!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Type of eventNullableInt is {eventNullableInt.GetType().Name}");
Assert.Equal(eventNullableInt, 12);
```

Local clearly shows it's an `Int32`:

data type - local

Was hoping to see this on the console output of the test run from the CI, but it's not showing up for whatever reason. That would have been more of a "smoking gun". Will have to settle for the exception info that shows it to be `StructValue`:

ci exception

Whatever's going on, I can't get it to repro locally. My next strategy is to make a sacrifice to CI to atone for my sins, because I've obviously angered it somehow. I've got another PR open where test summary shows 4 fails, but details says everything passed.

**Comment by jbhensley** [comment]
@dotnetbot test this please

**Comment by jbhensley** [comment]
@dotnet-bot test this please

**Comment by jbhensley** [comment]
@dotnet-bot test this please

**Comment by jbhensley** [comment]
Hmmm. CI no longer running. Maybe some kind of rate limit that I've exceeded. Current change will dump type name in the xUnit `Assert`. I know it's `System.Int32` on local.

**Comment by noahfalk** [comment]
You should be able to get output from xunit by doing this:
https://xunit.github.io/docs/capturing-output.html

I don't have a good theory about why you are getting different results local vs. remote, but I'm hoping being able to write to the test log will make debugging it a little more sane ; ) I agree that from the output you have it does appear that on the CI server eventNullableInt isn't a Nullable<int> type.

`var eventNullableInt = evt.PayloadValue(0, "nInteger");`

Hopefully the result of eventNullableInt.GetType().Name will be illuminating.

**Comment by jbhensley** [comment]
[This test](https://ci3.dot.net/job/dotnet_corefx/job/master/job/windows-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/17529/) had xUnit output turned on, but I'm not seeing a way to view the full test execution log.

One test asserted the namespace + type name, but cut off at `Microsoft.Diagnostics.Tracing.Parsers.Str···`. Another test gave the type name as `StructValue`. Assuming full namespace+type=`Microsoft.Diagnostics.Tracing.Parsers.StructValue`. This is an `internal` scoped class in PerfView located [here](https://github.com/Microsoft/perfview/blob/0f77be5938135af9a64c2f928c44b19dc7dbc048/src/TraceEvent/DynamicTraceEventParser.cs#L775). Not seeing anything in core that corresponds to it.

Working theory: CI is injecting its own parses to basically do its job and this is somehow causing a side-effect.

**Comment by jbhensley** [comment]
@vancem Any ideas on above?

**Comment by jbhensley** [comment]
Thanks, that makes sense. 

So what I'm seeing: if I add this code to our test harness

```C#
int? nullableInt = 12;
tests.Add(new SubTest("Write/Basic/int?/12",
	delegate ()
	{
		logger.Write("Int12", new { nInteger = nullableInt });
	},
	delegate (Event evt)
	{
		var payload = evt.PayloadValue(0, "nInteger");
		Assert.Equal(12, payload);
	}
```

This passes locally because the `type` of `payload` is `Int32` (locally it is `Int32`). On the CI, it fails because the `type` of `payload` is `Microsoft.Diagnostics.Tracing.Parsers.StructValue`. Following the lead of other code in the harness, I tried doing the following:

```C#
var structValueAsDictionary = payload as IDictionary<string, object>;
Assert.NotNull(structValueAsDictionary);
```

but that tossed at the `Assert`. 

So, I think the summary is: Locally, nullables are coming back as boxed values (either `Int32` or a null object reference). On the CI they are `Microsoft.Diagnostics.Tracing.Parsers.StructValue`. However, on the CI casting `Microsoft.Diagnostics.Tracing.Parsers.StructValue` to `IDictionary<string, object>` fails, so its not clear how to properly handle it.

One theory: could the CI be somehow globally registering for trace events, handling them as part of its work and then passing things it does not handle down a call chain, eventually ending up at our code? In that way, it might be altering the event before it reaches us.

**Comment by vancem** [comment]
I have a bit more information that may be of use.    You are updating the Test_Write_T method, but that method gets called in two very different modes.   It is called by the Test_Write_T_ETW and Test_Write_T_EventListener.   In this issue it is very important to understand which test is actually failing.  

The ETW version of the test forces the data sent 'out of process' and it uses a ETW library called TraceEvent to parse the resulting data.   The writing logic and the TraceEvent library has no special knowledge of Nullable, and thus serializes a two fields and reads it back as a StructValue.   However the Test_Write_T_EventListener version goes through a different serialization/deserialization path, and round drips it to a nullable, which it exposes as an object (which means it boxes it) which means it shows up as either null or the T (in this case int).   

Thus what I would have EXPECTED is that both local and in CI ONE of (Test_Write_T_ETW  and Test_Write_T_EventListener) would fail.   As mentioned I can't explain why it works, but lets debug what we can locally.     On your local machine when running Test_Write_T_ETW, you shoudl get a StructValue (which should fail locally we should find out why it succeeds for you), and when running Test_Write_T_EventListener you should get a boxed int.   

To fix the test for good, make up a function 'UnwrapNullable<T>(object)' which tests if it is null, and if not if it is an IDictionary<string, Object> if so, assert that it has a 'HasValue' and 'Value' field and then return the value field.   Otherwise just return the original object.   This will work in either case, and shoudl make the tests pass.

**Comment by jbhensley** [comment]
Yes! That it! I didn't recognize that it was two different scenarios failing. What compounded this is that the ETW test has to meet the `IsProcessElevatedAndNotWindowsNanoServer` condition, which it did not locally because my cmd is not elevated. So, locally only the `EventListener` test ran. Because of this, casting `payload` to `Int32` always passed locally. On the CI, the cast failed the ETW test. When I tried casting to `IDictionary`, it would fail local and on the CI, but I didn't realize that it was a completely separate scenario on the CI than the first time, so it just looked like the CI would fail no matter which I tried.

Now that I can test local (by running an elevated cmd), I'll get this fixed and check in the changes. Thanks for the help.

**Comment by vancem** [comment]
This looks good.  Thanks @jbhensley for persisting in this effort.   You had to push through more that you probably originally signed up for.   Thanks for your persistance!   Good job! and thanks again.

Vance

**Comment by jbhensley** [comment]
No problem. I think this clears up the original issue. 

I did note another error condition during my testing that we should look at, but I'll open a separate issue for it. For this test:

```C#
tests.Add(new SubTest("WriteEvent/SelfDescribingOnly/UserData",
delegate ()
{
	logger.EventUserDataInt(new UserData() { x = 3, y = 8 }, 5);
},
delegate (Event evt)
{
	Assert.Equal(logger.Name, evt.ProviderName);
	Assert.Equal("EventUserDataInt", evt.EventName);

	var aClass = (IDictionary<string, object>)evt.PayloadValue(0, "aClass");
	Assert.Equal(3, (int)aClass["x"]);
	Assert.Equal(8, (int)aClass["y"]);
	Assert.Equal(5, evt.PayloadValue(1, "anInt"));
}));
```

Modifying `UserData` by adding four nullables like so:

```C#
[EventData]
public class UserData
{
	public int x { get; set; }
	public int y { get; set; }

	/*
	 *      These lines result in assert fail at 417:
	 *          Assert.Equal(5, evt.PayloadValue(1, "anInt"));
	 */

	public int? N1 { get; set; } = int.MaxValue;
	public int? N2 { get; set; } = int.MaxValue;
	public int? N3 { get; set; } = int.MaxValue;
	public int? N4 { get; set; } = int.MaxValue;
}
```

results in invalid data coming back. If I dump the event, this is what I see:

```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
  Dump of Event
  Name EventUserDataInt
  Payload count: 2
  Payload:
  Name: aClass  Value: { x:3, y:8, N1:{ HasValue:True, Value:2147483647 }, N2:{ HasValue:True, Value:2147483647 }, N3:{ HasValue:True, Value:2147483647 } }
  Name: anInt   Value: -255
  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Note that the fourth nullable disappears out of the payload and we appear to stomp on the memory space of `anInt`, changing its value from 5 to -255. This results in an assert fail at `	Assert.Equal(5, evt.PayloadValue(1, "anInt"));`.

The concerning thing here is this looks suspiciously like a buffer overflow.

**Comment by vancem** [comment]
@jbhensley  - yes does look like a bug, thanks for filing it.   P.S:  FYI: picking test values that have different values in every byte of the int at the best values e.g. instead of MaxInt use 0x98765432  (and different ones in each field) to insure that any buffer issues (even of one byte) are more likely to be found (not passed 'just by accident).  

@noahfalk - feel free to merge at your convenience.

**Comment by jbhensley** [comment]
@vancem That's a good tip. I was originally packing the payload with two `int?` and two `DateTime?`, where one had a value and one was `null` for each. When the assert for `anInt` came back false, I noticed that changing the last `DateTime?` value would change the value of `anInt`. Then I just changed all to `int?` to see if `DateTime` itself was the problem.

I did try packing a bunch of non-nullable `int` and did not have a problem, so it looks like nullable is required to repro.

**Pull request merged**